### DISCUSSION TITLE
HOW to create a dockerfile for YTuner and Raspi

### DISCUSSION THREAD

**Discussion by qsm101** [coffeegreg/YTuner#40] [Q&A]
Hi all,
I am still struggeling - even with all the hints to create a dockerfile to run YTuner with Raspi.
Appreciate more details from the audience.

BR and thanks
  Juergen

**Comment by Larsvb0** [comment]
follow this and tell us when it fails:

# YTuner in Docker container
If you can't live without containers and you think that every day without creating a new container is a wasted day, you can also run YTuner in a container.
## Assumptions
In my opinion, there is no point in placing `cache`, `config` and `db` with the contents in the container and we should use `CacheFolderLocation`, `ConfigFolderLocation` and `DBFolderLocation` options of `ytuner.ini` file and point to host resources to remain persistent.
Therefore, in this short description I'll focus on this scenario, but of course everyone can choose their own.
>Tip: In the description below, I'll use Ubuntu 22.04 LTS as a host, `x86_64-linux` YTuner binaries and Alpine Linux Docker image to make my container as small as possible.

If you are familiar with building Docker containers, you may want to skip reading further. If not, don't worry and read on.
## That's easy
### 1) Install Docker
Use a well-documented procedure: https://docs.docker.com/engine/install/ubuntu/#install-using-the-repository
### 2) Prepare YTuner binaries and configuration
First, let's prepare some host shared volume (i.e. ~/YTuner) and copy your YTuner's config and `my_station.ini/yaml` files.
```
ladm@localubuntu20:~$ mkdir YTuner && cd YTuner && mkdir config
ladm@localubuntu20:~/YTuner$ 
```
Now, let's prepare `YTuner-docker` directory :
```
ladm@localubuntu20:~$ mkdir YTuner-docker && cd YTuner-docker
ladm@localubuntu20:~/YTuner-docker$ 
```
Get YTuner binaries :
```
ladm@localubuntu20:~/YTuner-docker$ wget https://github.com/coffeegreg/YTuner/releases/download/1.2.2/ytuner-1.2.2-x86_64-linux.zip
```
or build your own from source.

Extract `ytuner` and `ytuner.ini` files only :
```
ladm@localubuntu20:~/YTuner-docker$ unzip -d x86_64-linux/ ytuner-1.2.2-x86_64-linux.zip ytuner ytuner.ini
```
Make `ytuner` executable :
```
ladm@localubuntu20:~/YTuner-docker$ sudo chmod +x x86_64-linux/ytuner
```
Edit and change the `ytuner.ini` file as you wish, but remember to set the following options :
```
CacheFolderLocation=/app/host-shared
ConfigFolderLocation=/app/host-shared
DBFolderLocation=/app/host-shared
```
Now prepare the following `Dockerfile` :
```
# Start with the Alpine 3.19 Docker image.
FROM alpine:3.19

# Install libc6-compat, required for executables generated by FPC.
RUN apk --no-cache add libc6-compat
# Install sqlite-libs to support [catDB,catMemDB,catPermMemDB] cache types.
RUN apk --no-cache add sqlite-libs

# Set working directory in the container image.
WORKDIR /app

# Copy the YTuner files from host into container image.
COPY x86_64-linux/ /app/
 
# Make the TCP/UDP ports to outside the container.
EXPOSE 80/tcp
EXPOSE 53/udp

# Run YTuner.
CMD ["./ytuner"]
```
### 3) Building Docker container
Now you are ready to build your Docker container with YTuner binaries :
```
ladm@localubuntu20:~/YTuner-docker$ sudo docker build -t ytuner:1.2.2 .
+] Building 0.7s (10/10) FINISHED                                                                                                   docker:default
 => [internal] load build definition from Dockerfile                                                                                           0.0s
 => => transferring dockerfile: 674B                                                                                                           0.0s
 => [internal] load metadata for docker.io/library/alpine:3.19                                                                                 0.5s
 => [internal] load .dockerignore                                                                                                              0.0s
 => => transferring context: 2B                                                                                                                0.0s
 => [1/5] FROM docker.io/library/alpine:3.19@sha256:51b67269f354137895d43f3b3d810bfacd3945438e94dc5ac55fdac340352f48                           0.0s
 => [internal] load build context                                                                                                              0.0s
 => => transferring context: 118B                                                                                                              0.0s
 => CACHED [2/5] RUN apk --no-cache add libc6-compat                                                                                           0.0s
 => CACHED [3/5] RUN apk --no-cache add sqlite-libs                                                                                            0.0s
 => CACHED [4/5] WORKDIR /app                                                                                                                  0.0s
 => CACHED [5/5] COPY x86_64-linux/ /app/                                                                                                      0.0s
 => exporting to image                                                                                                                         0.0s
 => => exporting layers                                                                                                                        0.0s
 => => writing image sha256:0f5a6853d85c813d2351ff27199b2576b619d63863eb4020fa4e08feaf22f7e6                                                   0.0s
 => => naming to docker.io/library/ytuner:1.2.2                                                                                                0.0s
```
>Tip: 1.2.2 is a version related tag. You can use other values.

Congratulations, your YTuner Docker container has been built. You can see it :
```
ladm@localubuntu20:~/YTuner-docker$ sudo docker images
REPOSITORY    TAG       IMAGE ID       CREATED          SIZE
ytuner        1.2.2     0f5a6853d85c   26 seconds ago   11.4MB
```
### 4) Running YTuner in Docker container
Now, you're ready to go:
```
ladm@localubuntu20:~/YTuner-docker$ sudo docker run --rm --network host -v ~/YTuner:/app/host-shared ytuner:1.2.2
YTuner v1.2.2 Copyright (c) 2024 Greg P. (https://github.com/coffeegreg)
24-1-24 22:20:47 : Inf : Starting services...
24-1-24 22:20:47 : Inf : Successfully loaded 10 my stations.
24-1-24 22:20:47 : Dbg : Database library : 3.44.2.
24-1-24 22:20:47 : Inf : Checking local database.
24-1-24 22:20:47 : Inf : Preparing local database This may take a while...
24-1-24 22:20:47 : Dbg : Creating tables.
24-1-24 22:20:47 : Dbg : Filling out countries table.
24-1-24 22:20:47 : Dbg : Radio-browser.info: GET countries.
24-1-24 22:20:48 : Dbg : Filling out languages table.
24-1-24 22:20:48 : Dbg : Radio-browser.info: GET languages.
24-1-24 22:20:48 : Dbg : Filling out codecs table.
24-1-24 22:20:48 : Dbg : Radio-browser.info: GET codecs.
24-1-24 22:20:48 : Dbg : Filling out tags table.
24-1-24 22:20:48 : Dbg : Radio-browser.info: GET tags.
24-1-24 22:20:49 : Dbg : Creating indexes.
24-1-24 22:20:49 : Dbg : Filling out stations table.
24-1-24 22:20:49 : Dbg : Radio-browser.info: GET stations.
24-1-24 22:21:54 : Dbg : Cleaning tables.
24-1-24 22:22:19 : Dbg : Rebuilding indexes.
24-1-24 22:22:20 : Inf : Local database is ready.
24-1-24 22:22:20 : Inf : DNS Service: listening on: 192.168.1.140:53.
24-1-24 22:22:20 : Inf : Web Service: listening on: 192.168.1.140:80.
```
If you stop running YTuner container and start it again YTuner will use database or cache files already stored on host shared volume. 
```
ladm@localubuntu20:~/YTuner-docker$ sudo docker run --rm --network host -v ~/YTuner:/app/host-shared ytuner:1.2.2
YTuner v1.2.2 Copyright (c) 2024 Greg P. (https://github.com/coffeegreg)
24-1-24 22:17:38 : Inf : Starting services...
24-1-24 22:17:38 : Inf : Successfully loaded 10 my stations.
24-1-24 22:17:38 : Dbg : Database library : 3.44.2.
24-1-24 22:17:38 : Inf : Checking local database.
24-1-24 22:17:38 : Inf : Local database is ready.
24-1-24 22:17:38 : Inf : DNS Service: listening on: 192.168.1.140:53.
24-1-24 22:17:38 : Inf : Web Service: listening on: 192.168.1.140:80.
```
That's it. Easy, right?
## Summary
This is, of course, a simplified description and you can change the container configuration according to your needs.
Use the docker documentation to learn more (https://docs.docker.com).

If you found any errors or inaccuracies in this description, please let me know.

**Comment by coffeegreg** [comment]
Hi Juergen
Start by installing docker, follow this procedure : [https://docs.docker.com/engine/install/raspberry-pi-os/](https://docs.docker.com/engine/install/raspberry-pi-os/)
Then follow the tutorial : [DOCKER.md](https://github.com/coffeegreg/YTuner/blob/master/doc/DOCKER.md) (The same one that Larsvb0 mentioned).
Pay attention to the `Dockerfile` :
```
# Start with the Alpine 3.19 Docker image.
FROM alpine:3.19
```
If you do not want to build a container based on Alpine Linux, choose another one that suits your needs.
Remember that Alpine Linux (or any musl libc systems) will need :
```
# Install libc6-compat, required for executables generated by FPC.
RUN apk --no-cache add libc6-compat
```
Just follow the tutorial and everything should be fine ... :+1: 

If you have any problems, write specifically what they are related to. Add logs/console dumps and we'll see how we can help you.

**Comment by qsm101** [comment]
Coffeegreg, 
thanks for your answer.
I tried again your proposal. after executing:
sudo docker run --rm --network host -v ~/YTuner:/app/host-shared ytuner:1.2.2
I get:
docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: exec: "./ytuner": permission denied: unknown.

sudo docker images
shows me, that I have an ytuner image created.

What is my error?

BR
 Juergen

**Reply by TheBossME** [reply]
chmod +x ytuner

**Comment by qsm101** [comment]
Hi TheBossME,

I did that already earlier. See here please:
pi@BPI181:~/YTuner-docker$ ls -l x86_64-linux/
total 2352
-rwxrwxrwx 1 pi pi 2395240 24. Jan 2024  ytuner
-rw-r--r-- 1 pi pi    9828 22. Jan 2024  ytuner.ini

But error remains when exectuing the sudo docker run command:
docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: exec: "./ytuner": permission denied: unknown.

I am still think, that the problem is because I am running on a raspi.

**Reply by TheBossME** [reply]
Which binary have you tried ?. Keep in mind you should use Arm64 architecture if on arm64.

I think, you're going failed with architecture.

I'm using ytuner in Proxmox 8.x on raspi 4 in LXC Containers without issues. You can also start ytuner directly on raspi. Why using docker ?. Why easy, when it's possible to complicate things 🤣👍🫶

Please type:

uname -a in your raspi console and post output here.

**Reply by qsm101** [reply]
uname -a shows:
Linux BPI181 6.6.62-current-sunxi #1 SMP Sun Nov 17 14:08:59 UTC 2024 armv7l GNU/Linux

**Reply by TheBossME** [reply]
armv7l GNU/Linux = 32 Bit armhf Linux

Why did you try x86_amd64 binary to run ?

What kind of raspberry Hardware is used from your end ?

**Reply by qsm101** [reply]
the hw is a banana pi raspi 2 - pretty old
I used ytuner-1.2.3-arm-linux-rpi$ as well, but did not execute as within the example of the x86_amd64.
when I executed: sudo docker run --rm --network host -v ~/YTuner:/app/host-shared ytuner:1.2.2
it just ended with the cmd prompt on linux and that is it.

**Reply by qsm101** [reply]
If i use the same as a standalone, it ends with: 
pi@BPI181:~/YTuner-docker/ytuner-1.2.3-arm-linux-rpi$ ./ytuner
YTuner v1.2.3 Copyright (c) 2024 Greg P. (https://github.com/coffeegreg)
28-12-24 12:02:36 : Wrn : You are running out of INI file or your ytuner.ini file is outdated! Some features may not work properly! Check https://github.com/coffeegreg/YTuner/tree/master/cfg for the correct INI file for your version of YTuner.
28-12-24 12:02:36 : Inf : Starting services...
28-12-24 12:02:36 : Inf : Successfully loaded 10 my stations.
28-12-24 12:02:36 : Inf : Checking local database.
28-12-24 12:02:36 : Inf : Local database is ready.
28-12-24 12:02:37 : Inf : Web Service: listening on: 192.168.0.181:80.
Exception at 000C9674: ESocketError:
Binding of socket failed: 80.
free(): invalid pointer

**Reply by TheBossME** [reply]
Sorry to say. Please gave up that project, you didn't have enough Linux / Unix knowledge.

Port 80 is blocked by other service. You can try 
netstat or ss -tulpn to analyze wich application is using port 80

**Reply by qsm101** [reply]
ok, thx for your help!
Best wishes
  Juergen

**Reply by coffeegreg** [reply]
1. Use [ytuner-1.2.4-armhf-linux.zip](https://github.com/coffeegreg/YTuner/releases/download/1.2.4/ytuner-1.2.4-armhf-linux.zip) from Releases. 
2. Use default `ytuner.ini` file.
3. Try to run YTuner with sudo : `sudo ./ytuner`
4. Is there any service listening on TCP port 80 ? :thinking: 
If you use other web service listening on TCP port 80 you can run YTuner web service on another TCP port (i.e. 8080) but you have to redirect all AVR's request to TCP port 80 to YTuner web service TCP port (i.e. 8080).
Read carefully all comments from [ytuner.ini](https://github.com/coffeegreg/YTuner/blob/master/cfg/ytuner.ini) and [avr.ini](https://github.com/coffeegreg/YTuner/blob/master/cfg/avr.ini)

**Reply by TheBossME** [reply]
Jürgen, don't be angry with me. But you don't seem to have any idea what you're doing. 

I could help you start the project. But if you don't understand the little things and I still don't know what kind of Raspi with which OS you're using under the hood, it doesn't make any sense.

**Reply by qsm101** [reply]
hi TheBossME,
no worries, I am not angry with you. I am aware to miss quite a lot knowledge on linux and docker. 

I am using a Bananapi M2U / Berry.
Earlier, I was able to run: mpvd/vtuner-emulator-ycast on docker before trying on coffeegrep version of ytuner.

Port 80 is working and after removing and move back the the power cable of my RX-V500D, i can hear music. The version I used
was on a cmd line ytuner-1.2.3-arm-linux-rpi. The key enabler was the sudo before the ./ytuner.
So step 1 is done.
Now let me see, if it works with ytuner-1.2.4-armhf-linux.

**Reply by coffeegreg** [reply]
>  i can hear music

Great ! :+1:

> The key enabler was the sudo before the ./ytuner

:+1: 

> Now let me see, if it works with ytuner-1.2.4-armhf-linux.


`ytuner-1.2.4-armhf-linux` has been build based on the same RPi libraries as previous versions.
I removed the `-rpi` postfix and changed name because I thought everyone would know that for a 32bit RPi, `armhf` would be the right choice.  :thinking: I think I should add the same file to Release but with `-rpi` postfix in the name. :thinking:

**Reply by qsm101** [reply]
I can confirm, 1.2.4 with armhf-linux is of course working. the idea with the postfix I like. Easier for those with less experience. Thx for your addtional patience. 
No worries, I did not take me that many hours. I just made a break to do some weekend shopping.
Now let me see, if I can do that with docker as well.

**Reply by coffeegreg** [reply]
> .... after removing and move back the the power cable of my RX-V500D, i can hear music. 

@qsm101
May I add your RX-V500D to the "Confirmed working" list? :thinking:

**Reply by qsm101** [reply]
of course. Please do so.

**Comment by vlad-6502** [comment]
Managed to run Ytuner in Docker on Raspberry PI on which my Home Assistant instance is currently running .

The only problem I faced - had to stop  **systemd-resolve** service which was interfering with Ytuner, it can't bind 53 port.
I'm not sure if I did this in an elegant way, what do you think ?

P.S. Seems it's also make sense:
`sudo systemctl disable systemd-resolved`

```
vlad@raspberrypi:~ $ sudo fuser 53/tcp
53/tcp:              17691
vlad@raspberrypi:~ $ sudo ps -p 17691 -o comm=
systemd-resolve
vlad@raspberrypi:~ $ sudo systemctl stop systemd-resolved
vlad@raspberrypi:~ $ sudo docker run --rm --network host -v ~/YTuner:/app/host-shared ytuner:1.2.2

YTuner v1.2.6 Copyright (c) 2024 Greg P. (https://github.com/coffeegreg)
9-2-25 15:36:51 : Inf : Starting services...
9-2-25 15:36:51 : Inf : Successfully loaded 11 my stations.
9-2-25 15:36:51 : Inf : Loading RB UUIDs cache file.
9-2-25 15:36:51 : Inf : Successfully loaded 53310 RB UUIDs from cache file.
9-2-25 15:36:51 : Inf : DNS Service: listening on: 102.108.1.121:53.
9-2-25 15:36:51 : Inf : Web Service: listening on: 102.108.1.121:80.





```

**Reply by coffeegreg** [reply]
`sudo systemctl disable systemd-resolved` is :+1:

**Comment by NeoXTof** [comment]
Hello @coffeegreg, I'm trying to run ytuner in a docker container, not exactly as you described it.

For security reason, I don't want to run the docker container on the host network. So I'm running it on a docker network. Everything works fine except the DNS Intercept. 

Whatever IP I configure in the ytuner.ini, the DNS intercept returns the IP of the container in the docker network and not the IP of the host.

Could you please either add a new paramter for the DNS intercept (to configure the IP of the host) or use the one that can be configured as "ActAsHost" ?

Best regards
Neo

**Reply by coffeegreg** [reply]
Hi @NeoXTof 
Look here : https://github.com/coffeegreg/YTuner/issues/78

**Reply by NeoXTof** [reply]
I'm hosting my docker on a pi 5. If you will include the #78 fix in the next release, I'll wait until then (but I can validate if on Pi for you if you want)
